### PR TITLE
#1210 Stocktake lines for stock lines with 0 snapshot quantity 

### LINKED ISSUE #1210: Stocktake lines for stocklines with 0 snapshot quantity
## What went wrong? 😲

Stocktake lines for stocklines with 0 total quantity appear in stocktake AND it looks like in outbound shipment issue modal

## Expected behaviour 🤔

Stocklines with 0 total quantity should be filtered out when creating stocktake line or when creating draft outbound shipment lines.

## How to Reproduce 🔨

Make sure you have at least one stockline (batch) with 0 total quanity

Create stocktake or outbound shipment with that item

## Your environment 🌱
 - Development on develop branch

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #1210 

# 👩🏻‍💻 What does this PR do? 
Only load stock lines for stocktake and outbound shipments if stock lines have stock

# 🧪 How has/should this change been tested? 
- Have an item with multiple batches
- Have one or more of those batches with 0 available number of packs
- Try to create a stocktake or outbound shipment with said item

### CODE REVIEW

**Comment by andreievg** [comment]
@roxy-dao, is_available deals with available quantity

**Comment by roxy-dao** [comment]
> a31a987

**Comment by andreievg** [comment]
> Shouldn't users not see the stock line if there's no available number of packs (since you shouldn't be able to allocate more than you have left)...? 🤔 Changed to total anyways

> Edit: If you create an outbound shipment with stock line a and assign nearly all the packs (keep as new), then create a new outbound shipment with same item, it creates a placeholder line for said item... so makes sense to me to use available number of packs instead of total number of packs.....

Sorry I should have explained.

For Outbound Shipments, in mSupply we still show those. `Total` = hasn't left facility, `Not Available` = in facility but it's been allocated to a shipments. If there is an urgent shipment we can still send `Not Available` (be reducing it form other shipment) and having that visible when issuing stock I think is the reason it's shown in mSupply

**Review by mark-prins — DISMISSED** [review]
👍 changes look good thanks!
am wondering if it necessary to add a similar filter to the outbound shipment to catch the second part of the issue? There shouldn't be any 0 packs due to stocktakes now - and I'd hope that mSupply won't send us any stock lines with 0 quantity, so that's probably a theoretical problem rather than a real one

**Comment by mark-prins** [comment]
Found a problem sorry - if you create a stocktake using master list or location then the test isn't applied and you can add stock lines which have zero quantity.

**Review by mark-prins — CHANGES_REQUESTED** [review]
Could you also add the filter to the stocktake creation by master list and location 🙏

**Comment by roxy-dao** [comment]
> Could you also add the filter to the stocktake creation by master list and location 🙏

Yup, changing it right now to do frontend filter instead of be (: (need to update from develop first though 🤣 )

File: client/packages/common/src/types/schema.ts [hunk 1]
```diff
@@ -2964,6 +2965,11 @@ export type StoreNode = {
   __typename: 'StoreNode';
   code: Scalars['String'];
   id: Scalars['String'];
+  /**
```
Comment (roxy-dao) [inline client/packages/common/src/types/schema.ts:2968]: Not sure why this was generated now... didn't go through with the associated PR?

File: client/packages/system/src/Stock/api/api.ts [hunk 2]
```diff
@@ -70,13 +70,17 @@ export const getStockQueries = (stockApi: StockApi, storeId: string) => ({
       nodes: StockLineRowFragment[];
       totalCount: number;
     }> => {
+      const filter = {
```
Comment (roxy-dao) [inline client/packages/system/src/Stock/api/api.ts:73]: Andrei said to filter in FE... but Stocktake by Masterlist and Location are generated in BE so 🤷🏻

**Review by andreievg — APPROVED** [review]
thanks for changes. looks good, something that I would have done different but it's probably out of scope for the issue and not required for this to be merged:

> Andrei said to filter in FE... but Stocktake by Masterlist and Location are generated in BE so

They are but there is API for that, I would actually change API stocktake input to have an option:

```diff
input InsertStocktakeInput {
        description: String
        isLocked: Boolean
        stocktakeDate: NaiveDate
-       masterListId: String
-       locationId: String
+       addItems: StocktakeAddItems
+}
+
+input StocktakeAddItems {
+       from: StocktakeAddItemFrom!
+       id: String!
+       stocklineFilter: StockLineFilterInput
+}
+
+enum StocktakeAddItemFrom {
+       MASTER_LIST
+       LOCATION
 }
```

`stocklineFilter` would apply on top of masterlist/location filter. 

Also just noticed, that we usually get stock lines (outbound/stocktake edit line modal) from item.availableBatches, I think renaming to item.batches and  adding filter on the batch :

```diff
-    async fn available_batches(
+    async fn batches(
         &self,
         ctx: &Context<'_>,
         store_id: String,
+        filter: Option<StockLineFilter>,
```

(fyi, the loader would need to use payload on IdPair, similar to `ItemStatsForItemLoader`)

**Review by mark-prins — APPROVED** [review]

**Pull request merged**